tiled/adapters/zarr.py

```
@@ -176,13 +174,64 @@ async def write_block(
         -------
 
         """
-        if slice is not ...:
-            raise NotImplementedError
         block_slice, shape = slice_and_shape_from_block_and_chunks(
             block, self.structure().chunks
         )
         self._array[block_slice] = data
 
+    async def patch(
+        self,
+        data: NDArray[Any],
+        slice: Tuple[slice | int, ...],
+        extend: bool = False,
+    ) -> Tuple[Tuple[int, ...], Tuple[Tuple[int, ...], ...]]:
+        """
+        Write data into a slice of the array, maybe extending it.
+
+        If the specified slice does not fit into the array, and extend=True, the
+        array will be resized (expanded, never shrunk) to fit it.
+
+        Parameters
+        ----------
+        data : array-like
+        slice :
+            Where to place the new data
+        extend : bool
+            If slice does not fit wholly within the shape of the existing array,
+            reshape (expand) it to fit if this is True.
+
+        Raises
+        ------
+        ValueError :
+            If slice does not fit wholly with the shape of the existing array
+            and expand is False
+        """
+        current_shape = self._array.shape
+        new_shape = list(current_shape)
+        for i, (s, dim) in enumerate(zip(slice, current_shape)):
+            if isinstance(s, int):
+                new_shape[i] = max(new_shape[i], s)
+            elif isinstance(s, builtins.slice) and isinstance(s.stop, int):
+                new_shape[i] = max(new_shape[i], s.stop)
```

Should we raise an error if these conditions aren't met?